File: telemetry/aitf/sdk/go/instrumentation/asset_inventory.go
// Package instrumentation provides AITF instrumentation for AI operations.
package instrumentation

import (
	"context"
	"fmt"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/trace"

	"github.com/girdav01/AITF/sdk/go/semconv"
)

const assetInventoryTracerName = "instrumentation.asset_inventory"

// AssetInventoryInstrumentor traces AI asset lifecycle management including
// registration, discovery, audit, risk classification, and decommissioning.
// Aligned with CoSAI AI Incident Response preparation requirements.
type AssetInventoryInstrumentor struct {
	tracer trace.Tracer
}

// NewAssetInventoryInstrumentor creates a new asset inventory instrumentor.
func NewAssetInventoryInstrumentor(tp trace.TracerProvider) *AssetInventoryInstrumentor {
	if tp == nil {
		tp = otel.GetTracerProvider()
	}
	return &AssetInventoryInstrumentor{
		tracer: tp.Tracer(assetInventoryTracerName),
	}
}

// ── Registration ────────────────────────────────────────────────────────────

// RegisterConfig holds configuration for an asset registration span.
type RegisterConfig struct {
	AssetID              string
	AssetName            string
	AssetType            string // e.g. "model", "dataset", "agent", "pipeline"
	Version              string
	AssetHash            string
	Owner                string
	OwnerType            string
	DeploymentEnvironment string
	RiskClassification   string
	SourceRepository     string
	Tags                 []string
}

// AssetRegistration provides methods for enriching an asset registration span.
type AssetRegistration struct {
	span    trace.Span
	AssetID string
}

// TraceRegister starts a span for asset registration. The caller must call
// End() when done.
func (ai *AssetInventoryInstrumentor) TraceRegister(ctx context.Context, cfg RegisterConfig) (context.Context, *AssetRegistration) {
	if cfg.AssetType == "" {
		cfg.AssetType = "model"
	}

	spanName := fmt.Sprintf("asset.register %s %s", cfg.AssetType, cfg.AssetName)
	attrs := []attribute.KeyValue{
		semconv.AssetIDKey.String(cfg.AssetID),
		semconv.AssetNameKey.String(cfg.AssetName),
		semconv.AssetTypeKey.String(cfg.AssetType),
	}
	if cfg.Version != "" {
		attrs = append(attrs, semconv.AssetVersionKey.String(cfg.Version))
	}
	if cfg.AssetHash != "" {
		attrs = append(attrs, semconv.AssetHashKey.String(cfg.AssetHash))
	}
	if cfg.Owner != "" {
		attrs = append(attrs, semconv.AssetOwnerKey.String(cfg.Owner))
	}
	if cfg.OwnerType != "" {
		attrs = append(attrs, semconv.AssetOwnerTypeKey.String(cfg.OwnerType))
	}
	if cfg.DeploymentEnvironment != "" {
		attrs = append(attrs, semconv.AssetDeployEnvKey.String(cfg.DeploymentEnvironment))
	}
	if cfg.RiskClassification != "" {
		attrs = append(attrs, semconv.AssetRiskClassificationKey.String(cfg.RiskClassification))
	}
	if cfg.SourceRepository != "" {
		attrs = append(attrs, semconv.AssetSourceRepoKey.String(cfg.SourceRepository))
	}
	if len(cfg.Tags) > 0 {
		attrs = append(attrs, semconv.AssetTagsKey.StringSlice(cfg.Tags))
	}

	ctx, span := ai.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, &AssetRegistration{span: span, AssetID: cfg.AssetID}
}

// SetHash records the asset hash after registration.
func (r *AssetRegistration) SetHash(hash string) {
	r.span.SetAttributes(semconv.AssetHashKey.String(hash))
}

// SetVersion records the asset version.
func (r *AssetRegistration) SetVersion(version string) {
	r.span.SetAttributes(semconv.AssetVersionKey.String(version))
}

// SetRiskClassification records or updates the risk classification.
func (r *AssetRegistration) SetRiskClassification(classification string) {
	r.span.SetAttributes(semconv.AssetRiskClassificationKey.String(classification))
}

// SetDeploymentEnvironment records the deployment environment.
func (r *AssetRegistration) SetDeploymentEnvironment(env string) {
	r.span.SetAttributes(semconv.AssetDeployEnvKey.String(env))
}

// End completes the registration span.
func (r *AssetRegistration) End(err error) {
	if err != nil {
		r.span.SetStatus(codes.Error, err.Error())
		r.span.RecordError(err)
	} else {
		r.span.SetStatus(codes.Ok, "")
	}
	r.span.End()
}

// Span returns the underlying OTel span.
func (r *AssetRegistration) Span() trace.Span { return r.span }

// ── Discovery ───────────────────────────────────────────────────────────────

// DiscoveryConfig holds configuration for an asset discovery span.
type DiscoveryConfig struct {
	Scope  string // e.g. "organization", "team", "project"
	Method string // e.g. "api_scan", "network_scan", "registry_sync"
}

// AssetDiscovery provides methods for enriching an asset discovery span.
type AssetDiscovery struct {
	span trace.Span
}

// TraceDiscover starts a span for an asset discovery scan. The caller must call
// End() when done.
func (ai *AssetInventoryInstrumentor) TraceDiscover(ctx context.Context, cfg DiscoveryConfig) (context.Context, *AssetDiscovery) {
	if cfg.Scope == "" {
		cfg.Scope = "organization"
	}
	if cfg.Method == "" {
		cfg.Method = "api_scan"
	}

	spanName := fmt.Sprintf("asset.discover %s", cfg.Scope)
	attrs := []attribute.KeyValue{
		semconv.AssetDiscoveryScopeKey.String(cfg.Scope),
		semconv.AssetDiscoveryMethodKey.String(cfg.Method),
	}

	ctx, span := ai.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, &AssetDiscovery{span: span}
}

// SetResults records discovery scan results.
func (d *AssetDiscovery) SetResults(assetsFound, newAssets, shadowAssets int) {
	d.span.SetAttributes(
		semconv.AssetDiscoveryAssetsFoundKey.Int(assetsFound),
		semconv.AssetDiscoveryNewAssetsKey.Int(newAssets),
		semconv.AssetDiscoveryShadowKey.Int(shadowAssets),
	)
}

// SetStatus records the discovery status.
func (d *AssetDiscovery) SetStatus(status string) {
	d.span.SetAttributes(semconv.AssetDiscoveryStatusKey.String(status))
}

// End completes the discovery span.
func (d *AssetDiscovery) End(err error) {
	if err != nil {
		d.span.SetStatus(codes.Error, err.Error())
		d.span.RecordError(err)
	} else {
		d.span.SetStatus(codes.Ok, "")
	}
	d.span.End()
}

// Span returns the underlying OTel span.
func (d *AssetDiscovery) Span() trace.Span { return d.span }

// ── Audit ───────────────────────────────────────────────────────────────────

// AuditConfig holds configuration for an asset audit span.
type AuditConfig struct {
	AssetID   string
	AuditType string // e.g. "compliance", "security", "integrity"
	Framework string
	Auditor   string
}

// AssetAudit provides methods for enriching an asset audit span.
type AssetAudit struct {
	span trace.Span
}

// TraceAudit starts a span for an asset audit. The caller must call End() when done.
func (ai *AssetInventoryInstrumentor) TraceAudit(ctx context.Context, cfg AuditConfig) (context.Context, *AssetAudit) {
	if cfg.AuditType == "" {
		cfg.AuditType = "compliance"
	}

	spanName := fmt.Sprintf("asset.audit %s", cfg.AssetID)
	attrs := []attribute.KeyValue{
		semconv.AssetIDKey.String(cfg.AssetID),
		semconv.AssetAuditTypeKey.String(cfg.AuditType),
	}
	if cfg.Framework != "" {
		attrs = append(attrs, semconv.AssetAuditFrameworkKey.String(cfg.Framework))
	}
	if cfg.Auditor != "" {
		attrs = append(attrs, semconv.AssetAuditAuditorKey.String(cfg.Auditor))
	}

	ctx, span := ai.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, &AssetAudit{span: span}
}

// SetResult records the audit result.
func (a *AssetAudit) SetResult(result string) {
	a.span.SetAttributes(semconv.AssetAuditResultKey.String(result))
}

// SetRiskScore records the audit risk score.
func (a *AssetAudit) SetRiskScore(score float64) {
	a.span.SetAttributes(semconv.AssetAuditRiskScoreKey.Float64(score))
}

// SetIntegrityVerified records whether integrity was verified.
func (a *AssetAudit) SetIntegrityVerified(verified bool) {
	a.span.SetAttributes(semconv.AssetAuditIntegrityKey.Bool(verified))
}

// SetComplianceStatus records the compliance status.
func (a *AssetAudit) SetComplianceStatus(status string) {
	a.span.SetAttributes(semconv.AssetAuditComplianceKey.String(status))
}

// SetFindings records audit findings as a string.
func (a *AssetAudit) SetFindings(findings string) {
	a.span.SetAttributes(semconv.AssetAuditFindingsKey.String(findings))
}

// SetNextAuditDue records when the next audit is due.
func (a *AssetAudit) SetNextAuditDue(timestamp string) {
	a.span.SetAttributes(semconv.AssetAuditNextDueKey.String(timestamp))
}

// End completes the audit span.
func (a *AssetAudit) End(err error) {
	if err != nil {
		a.span.SetStatus(codes.Error, err.Error())
		a.span.RecordError(err)
	} else {
		a.span.SetStatus(codes.Ok, "")
	}
	a.span.End()
}

// Span returns the underlying OTel span.
func (a *AssetAudit) Span() trace.Span { return a.span }

// ── Classification ──────────────────────────────────────────────────────────

// ClassifyConfig holds configuration for a risk classification span.
type ClassifyConfig struct {
	AssetID            string
	RiskClassification string // e.g. "high_risk", "minimal_risk"
	Framework          string // e.g. "eu_ai_act"
	Assessor           string
	UseCase            string
}

// AssetClassification provides methods for enriching a risk classification span.
type AssetClassification struct {
	span trace.Span
}

// TraceClassify starts a span for risk classification. The caller must call
// End() when done.
func (ai *AssetInventoryInstrumentor) TraceClassify(ctx context.Context, cfg ClassifyConfig) (context.Context, *AssetClassification) {
	if cfg.Framework == "" {
		cfg.Framework = "eu_ai_act"
	}

	spanName := fmt.Sprintf("asset.classify %s", cfg.AssetID)
	attrs := []attribute.KeyValue{
		semconv.AssetIDKey.String(cfg.AssetID),
		semconv.AssetRiskClassificationKey.String(cfg.RiskClassification),
		semconv.AssetClassFrameworkKey.String(cfg.Framework),
	}
	if cfg.Assessor != "" {
		attrs = append(attrs, semconv.AssetClassAssessorKey.String(cfg.Assessor))
	}
	if cfg.UseCase != "" {
		attrs = append(attrs, semconv.AssetClassUseCaseKey.String(cfg.UseCase))
	}

	ctx, span := ai.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, &AssetClassification{span: span}
}

// SetPrevious records the previous risk classification.
func (c *AssetClassification) SetPrevious(previous string) {
	c.span.SetAttributes(semconv.AssetClassPreviousKey.String(previous))
}

// SetReason records the reason for classification.
func (c *AssetClassification) SetReason(reason string) {
	c.span.SetAttributes(semconv.AssetClassReasonKey.String(reason))
}

// SetAutonomousDecision records whether the asset makes autonomous decisions.
func (c *AssetClassification) SetAutonomousDecision(autonomous bool) {
	c.span.SetAttributes(semconv.AssetClassAutonomousDecisionKey.Bool(autonomous))
}

// End completes the classification span.
func (c *AssetClassification) End(err error) {
	if err != nil {
		c.span.SetStatus(codes.Error, err.Error())
		c.span.RecordError(err)
	} else {
		c.span.SetStatus(codes.Ok, "")
	}
	c.span.End()
}

// Span returns the underlying OTel span.
func (c *AssetClassification) Span() trace.Span { return c.span }

// ── Decommission ────────────────────────────────────────────────────────────

// DecommissionConfig holds configuration for an asset decommission span.
type DecommissionConfig struct {
	AssetID       string
	AssetType     string
	Reason        string
	ReplacementID string
	ApprovedBy    string
}

// TraceDecommission starts a span for asset decommissioning. The caller must
// call End() on the returned span when done.
func (ai *AssetInventoryInstrumentor) TraceDecommission(ctx context.Context, cfg DecommissionConfig) (context.Context, trace.Span) {
	spanName := fmt.Sprintf("asset.decommission %s %s", cfg.AssetType, cfg.AssetID)
	attrs := []attribute.KeyValue{
		semconv.AssetIDKey.String(cfg.AssetID),
		semconv.AssetTypeKey.String(cfg.AssetType),
		semconv.AssetDecommissionReasonKey.String(cfg.Reason),
	}
	if cfg.ReplacementID != "" {
		attrs = append(attrs, semconv.AssetDecommissionReplacementKey.String(cfg.ReplacementID))
	}
	if cfg.ApprovedBy != "" {
		attrs = append(attrs, semconv.AssetDecommissionApprovedByKey.String(cfg.ApprovedBy))
	}

	ctx, span := ai.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, span
}

File: telemetry/aitf/sdk/typescript/src/instrumentation/asset-inventory.ts
/**
 * AITF AI Asset Inventory Instrumentation.
 *
 * Provides tracing for AI asset lifecycle management: registration, discovery,
 * audit, risk classification, dependency mapping, and decommissioning. Aligned
 * with CoSAI AI Incident Response preparation requirements.
 */

import {
  trace,
  Span,
  SpanKind,
  SpanStatusCode,
  Tracer,
  TracerProvider,
} from "@opentelemetry/api";
import { AssetInventoryAttributes } from "../semantic-conventions/attributes";

const TRACER_NAME = "instrumentation.asset_inventory";

// ---------------------------------------------------------------------------
// Option interfaces
// ---------------------------------------------------------------------------

/** Options for tracing asset registration. */
export interface TraceRegisterOptions {
  assetId?: string;
  assetName?: string;
  assetType?: string;
  version?: string;
  assetHash?: string;
  owner?: string;
  ownerType?: string;
  deploymentEnvironment?: string;
  riskClassification?: string;
  sourceRepository?: string;
  tags?: string[];
}

/** Options for tracing asset discovery scans. */
export interface TraceDiscoverOptions {
  scope?: string;
  method?: string;
}

/** Options for tracing asset audits. */
export interface TraceAuditOptions {
  assetId: string;
  auditType?: string;
  framework?: string;
  auditor?: string;
}

/** Options for tracing risk classification. */
export interface TraceClassifyOptions {
  assetId: string;
  riskClassification: string;
  framework?: string;
  assessor?: string;
  useCase?: string;
}

/** Options for tracing asset decommissioning. */
export interface TraceDecommissionOptions {
  assetId: string;
  assetType: string;
  reason: string;
  replacementId?: string;
  approvedBy?: string;
}

// ---------------------------------------------------------------------------
// Helper classes
// ---------------------------------------------------------------------------

/** Helper for enriching an asset registration span. */
export class AssetRegistration {
  private readonly _span: Span;

  /** The unique identifier for this asset. */
  readonly assetId: string;

  constructor(span: Span, assetId: string) {
    this._span = span;
    this.assetId = assetId;
  }

  /** Access the underlying OTel span. */
  get span(): Span {
    return this._span;
  }

  /** Record the asset content hash. */
  setHash(hashValue: string): void {
    this._span.setAttribute(AssetInventoryAttributes.HASH, hashValue);
  }

  /** Record the asset version. */
  setVersion(version: string): void {
    this._span.setAttribute(AssetInventoryAttributes.VERSION, version);
  }

  /** Record the risk classification. */
  setRiskClassification(classification: string): void {
    this._span.setAttribute(AssetInventoryAttributes.RISK_CLASSIFICATION, classification);
  }

  /** Record the deployment environment. */
  setDeploymentEnvironment(env: string): void {
    this._span.setAttribute(AssetInventoryAttributes.DEPLOYMENT_ENVIRONMENT, env);
  }
}

/** Helper for enriching an asset discovery span. */
export class AssetDiscovery {
  private readonly _span: Span;

  constructor(span: Span) {
    this._span = span;
  }

  /** Access the underlying OTel span. */
  get span(): Span {
    return this._span;
  }

  /** Record discovery results. */
  setResults(options: {
    assetsFound: number;
    newAssets?: number;
    shadowAssets?: number;
  }): void {
    this._span.setAttribute(
      AssetInventoryAttributes.DISCOVERY_ASSETS_FOUND,
      options.assetsFound,
    );
    this._span.setAttribute(
      AssetInventoryAttributes.DISCOVERY_NEW_ASSETS,
      options.newAssets ?? 0,
    );
    this._span.setAttribute(
      AssetInventoryAttributes.DISCOVERY_SHADOW_ASSETS,
      options.shadowAssets ?? 0,
    );
  }

  /** Record the discovery status. */
  setStatus(status: string): void {
    this._span.setAttribute(AssetInventoryAttributes.DISCOVERY_STATUS, status);
  }
}

/** Helper for enriching an asset audit span. */
export class AssetAudit {
  private readonly _span: Span;

  constructor(span: Span) {
    this._span = span;
  }

  /** Access the underlying OTel span. */
  get span(): Span {
    return this._span;
  }

  /** Record the audit result. */
  setResult(result: string): void {
    this._span.setAttribute(AssetInventoryAttributes.AUDIT_RESULT, result);
  }

  /** Record the audit risk score. */
  setRiskScore(score: number): void {
    this._span.setAttribute(AssetInventoryAttributes.AUDIT_RISK_SCORE, score);
  }

  /** Record whether integrity was verified. */
  setIntegrityVerified(verified: boolean): void {
    this._span.setAttribute(AssetInventoryAttributes.AUDIT_INTEGRITY_VERIFIED, verified);
  }

  /** Record the compliance status. */
  setComplianceStatus(status: string): void {
    this._span.setAttribute(AssetInventoryAttributes.AUDIT_COMPLIANCE_STATUS, status);
  }

  /** Record audit findings as a string. */
  setFindings(findings: string): void {
    this._span.setAttribute(AssetInventoryAttributes.AUDIT_FINDINGS, findings);
  }

  /** Record when the next audit is due. */
  setNextAuditDue(timestamp: string): void {
    this._span.setAttribute(AssetInventoryAttributes.AUDIT_NEXT_AUDIT_DUE, timestamp);
  }
}

/** Helper for enriching a risk classification span. */
export class AssetClassification {
  private readonly _span: Span;

  constructor(span: Span) {
    this._span = span;
  }

  /** Access the underlying OTel span. */
  get span(): Span {
    return this._span;
  }

  /** Record the previous risk classification. */
  setPrevious(previous: string): void {
    this._span.setAttribute(AssetInventoryAttributes.CLASSIFICATION_PREVIOUS, previous);
  }

  /** Record the reason for the classification. */
  setReason(reason: string): void {
    this._span.setAttribute(AssetInventoryAttributes.CLASSIFICATION_REASON, reason);
  }

  /** Record whether the asset makes autonomous decisions. */
  setAutonomousDecision(autonomous: boolean): void {
    this._span.setAttribute(AssetInventoryAttributes.CLASSIFICATION_AUTONOMOUS_DECISION, autonomous);
  }
}

// ---------------------------------------------------------------------------
// Instrumentor
// ---------------------------------------------------------------------------

/**
 * Instrumentor for AI asset inventory operations.
 *
 * Traces asset registration, discovery, audit, risk classification, and
 * decommissioning with AITF semantic convention attributes.
 */
export class AssetInventoryInstrumentor {
  private _tracerProvider: TracerProvider | null;
  private _tracer: Tracer | null = null;
  private _instrumented = false;

  constructor(tracerProvider?: TracerProvider) {
    this._tracerProvider = tracerProvider ?? null;
  }

  /** Enable asset inventory instrumentation. */
  instrument(): void {
    const tp = this._tracerProvider ?? trace.getTracerProvider();
    this._tracer = tp.getTracer(TRACER_NAME);
    this._instrumented = true;
  }

  /** Disable asset inventory instrumentation. */
  uninstrument(): void {
    this._tracer = null;
    this._instrumented = false;
  }

  /** Get the tracer instance, lazily initializing if needed. */
  getTracer(): Tracer {
    if (!this._tracer) {
      const tp = this._tracerProvider ?? trace.getTracerProvider();
      this._tracer = tp.getTracer(TRACER_NAME);
    }
    return this._tracer;
  }

  // -- Registration ---------------------------------------------------------

  /**
   * Trace asset registration.
   *
   * Returns an {@link AssetRegistration} for manual span management, or
   * accepts a callback that receives the registration and automatically
   * ends the span.
   */
  traceRegister(options?: TraceRegisterOptions): AssetRegistration;
  traceRegister<T>(
    options: TraceRegisterOptions | undefined,
    fn: (reg: AssetRegistration) => T,
  ): T;
  traceRegister<T>(
    options?: TraceRegisterOptions,
    fn?: (reg: AssetRegistration) => T,
  ): AssetRegistration | T {
    const tracer = this.getTracer();
    const assetId = options?.assetId ?? crypto.randomUUID();
    const assetName = options?.assetName ?? "";
    const assetType = options?.assetType ?? "model";

    const attributes: Record<string, string | number | boolean | string[]> = {
      [AssetInventoryAttributes.ID]: assetId,
      [AssetInventoryAttributes.NAME]: assetName,
      [AssetInventoryAttributes.TYPE]: assetType,
    };
    if (options?.version !== undefined) {
      attributes[AssetInventoryAttributes.VERSION] = options.version;
    }
    if (options?.assetHash !== undefined) {
      attributes[AssetInventoryAttributes.HASH] = options.assetHash;
    }
    if (options?.owner !== undefined) {
      attributes[AssetInventoryAttributes.OWNER] = options.owner;
    }
    if (options?.ownerType !== undefined) {
      attributes[AssetInventoryAttributes.OWNER_TYPE] = options.ownerType;
    }
    if (options?.deploymentEnvironment !== undefined) {
      attributes[AssetInventoryAttributes.DEPLOYMENT_ENVIRONMENT] = options.deploymentEnvironment;
    }
    if (options?.riskClassification !== undefined) {
      attributes[AssetInventoryAttributes.RISK_CLASSIFICATION] = options.riskClassification;
    }
    if (options?.sourceRepository !== undefined) {
      attributes[AssetInventoryAttributes.SOURCE_REPOSITORY] = options.sourceRepository;
    }
    if (options?.tags !== undefined) {
      attributes[AssetInventoryAttributes.TAGS] = options.tags;
    }

    const spanName = `asset.register ${assetType} ${assetName}`;

    if (fn) {
      return tracer.startActiveSpan(
        spanName,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const reg = new AssetRegistration(otelSpan, assetId);
          try {
            const result = fn(reg);
            if (result instanceof Promise) {
              return result
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        },
      );
    }

    const otelSpan = tracer.startSpan(spanName, {
      kind: SpanKind.INTERNAL,
      attributes,
    });
    return new AssetRegistration(otelSpan, assetId);
  }

  // -- Discovery ------------------------------------------------------------

  /**
   * Trace asset discovery scans.
   *
   * Returns an {@link AssetDiscovery} for manual span management, or
   * accepts a callback that receives the discovery and automatically
   * ends the span.
   */
  traceDiscover(options?: TraceDiscoverOptions): AssetDiscovery;
  traceDiscover<T>(
    options: TraceDiscoverOptions | undefined,
    fn: (disc: AssetDiscovery) => T,
  ): T;
  traceDiscover<T>(
    options?: TraceDiscoverOptions,
    fn?: (disc: AssetDiscovery) => T,
  ): AssetDiscovery | T {
    const tracer = this.getTracer();
    const scope = options?.scope ?? "organization";
    const method = options?.method ?? "api_scan";

    const attributes: Record<string, string | number | boolean> = {
      [AssetInventoryAttributes.DISCOVERY_SCOPE]: scope,
      [AssetInventoryAttributes.DISCOVERY_METHOD]: method,
    };

    const spanName = `asset.discover ${scope}`;

    if (fn) {
      return tracer.startActiveSpan(
        spanName,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const disc = new AssetDiscovery(otelSpan);
          try {
            const result = fn(disc);
            if (result instanceof Promise) {
              return result
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        },
      );
    }

    const otelSpan = tracer.startSpan(spanName, {
      kind: SpanKind.INTERNAL,
      attributes,
    });
    return new AssetDiscovery(otelSpan);
  }

  // -- Audit ----------------------------------------------------------------

  /**
   * Trace asset audits.
   *
   * Returns an {@link AssetAudit} for manual span management, or accepts
   * a callback that receives the audit and automatically ends the span.
   */
  traceAudit(options: TraceAuditOptions): AssetAudit;
  traceAudit<T>(
    options: TraceAuditOptions,
    fn: (audit: AssetAudit) => T,
  ): T;
  traceAudit<T>(
    options: TraceAuditOptions,
    fn?: (audit: AssetAudit) => T,
  ): AssetAudit | T {
    const tracer = this.getTracer();
    const auditType = options.auditType ?? "compliance";

    const attributes: Record<string, string | number | boolean> = {
      [AssetInventoryAttributes.ID]: options.assetId,
      [AssetInventoryAttributes.AUDIT_TYPE]: auditType,
    };
    if (options.framework !== undefined) {
      attributes[AssetInventoryAttributes.AUDIT_FRAMEWORK] = options.framework;
    }
    if (options.auditor !== undefined) {
      attributes[AssetInventoryAttributes.AUDIT_AUDITOR] = options.auditor;
    }

    const spanName = `asset.audit ${options.assetId}`;

    if (fn) {
      return tracer.startActiveSpan(
        spanName,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const audit = new AssetAudit(otelSpan);
          try {
            const result = fn(audit);
            if (result instanceof Promise) {
              return result
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        },
      );
    }

    const otelSpan = tracer.startSpan(spanName, {
      kind: SpanKind.INTERNAL,
      attributes,
    });
    return new AssetAudit(otelSpan);
  }

  // -- Classification -------------------------------------------------------

  /**
   * Trace risk classification of an asset.
   *
   * Returns an {@link AssetClassification} for manual span management, or
   * accepts a callback that receives the classification and automatically
   * ends the span.
   */
  traceClassify(options: TraceClassifyOptions): AssetClassification;
  traceClassify<T>(
    options: TraceClassifyOptions,
    fn: (cls: AssetClassification) => T,
  ): T;
  traceClassify<T>(
    options: TraceClassifyOptions,
    fn?: (cls: AssetClassification) => T,
  ): AssetClassification | T {
    const tracer = this.getTracer();
    const framework = options.framework ?? "eu_ai_act";

    const attributes: Record<string, string | number | boolean> = {
      [AssetInventoryAttributes.ID]: options.assetId,
      [AssetInventoryAttributes.RISK_CLASSIFICATION]: options.riskClassification,
      [AssetInventoryAttributes.CLASSIFICATION_FRAMEWORK]: framework,
    };
    if (options.assessor !== undefined) {
      attributes[AssetInventoryAttributes.CLASSIFICATION_ASSESSOR] = options.assessor;
    }
    if (options.useCase !== undefined) {
      attributes[AssetInventoryAttributes.CLASSIFICATION_USE_CASE] = options.useCase;
    }

    const spanName = `asset.classify ${options.assetId}`;

    if (fn) {
      return tracer.startActiveSpan(
        spanName,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const cls = new AssetClassification(otelSpan);
          try {
            const result = fn(cls);
            if (result instanceof Promise) {
              return result
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        },
      );
    }

    const otelSpan = tracer.startSpan(spanName, {
      kind: SpanKind.INTERNAL,
      attributes,
    });
    return new AssetClassification(otelSpan);
  }

  // -- Decommission ---------------------------------------------------------

  /**
   * Trace asset decommissioning.
   *
   * Returns the underlying {@link Span} for manual management, or accepts
   * a callback that automatically ends the span.
   */
  traceDecommission(options: TraceDecommissionOptions): Span;
  traceDecommission<T>(
    options: TraceDecommissionOptions,
    fn: (span: Span) => T,
  ): T;
  traceDecommission<T>(
    options: TraceDecommissionOptions,
    fn?: (span: Span) => T,
  ): Span | T {
    const tracer = this.getTracer();

    const attributes: Record<string, string | number | boolean> = {
      [AssetInventoryAttributes.ID]: options.assetId,
      [AssetInventoryAttributes.TYPE]: options.assetType,
      [AssetInventoryAttributes.DECOMMISSION_REASON]: options.reason,
    };
    if (options.replacementId !== undefined) {
      attributes[AssetInventoryAttributes.DECOMMISSION_REPLACEMENT_ID] = options.replacementId;
    }
    if (options.approvedBy !== undefined) {
      attributes[AssetInventoryAttributes.DECOMMISSION_APPROVED_BY] = options.approvedBy;
    }

    const spanName = `asset.decommission ${options.assetType} ${options.assetId}`;

    if (fn) {
      return tracer.startActiveSpan(
        spanName,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          try {
            const result = fn(otelSpan);
            if (result instanceof Promise) {
              return result
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        },
      );
    }

    return tracer.startSpan(spanName, {
      kind: SpanKind.INTERNAL,
      attributes,
    });
  }
}
